File: 서인/CH06/1_올바른_괄호.js
function solution(s){
    let answer="YES";
    stack=[];

    if (s[0] === ")") return answer = "NO"

    for (let x of s){
      if (x === "(") stack.push(x)
      if (x === ")") stack.pop(x)
    }
    if (stack.length !== 0) {
      answer = "NO"
    }
     
    return answer;
}

File: 종복/Java/Stack&Queue/1. 올바른 괄호.java
import java.util.*;
class Main {	
	public String solution(String str){
		String answer="YES";
        Stack<Character> stk = new Stack<>();
      
        for(int i=0; i<str.length(); i++){
            if(str.charAt(i) == '('){
               stk.push('('); 
            }else{
              if(stk.size() == 0){
                  return "NO"; 
              }else{
                 stk.pop();
              }
            }
        }
      
        if(stk.size() != 0){
            return "NO";
        }
      
		return answer;
	}

	public static void main(String[] args){
		Main T = new Main();
		Scanner kb = new Scanner(System.in);
		String str=kb.next();
		System.out.println(T.solution(str));
	}
}
